finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
-                            *proposed_block.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
-                        }
-                    }
+    pub(crate) fn new(
+        select_chain: SC,
+        client: Arc<C>,
+        min_block_num: NumberFor<B>,
+        max_block_num: NumberFor<B>,
+        metrics: Option<Metrics<<B::Header as HeaderT>::Hash>>,
+    ) -> (Self, DataProvider<B>) {
+        let initial_data = AlephData::Empty;
+        let data_to_propose = Arc::new(Mutex::new(initial_data));
+        (
+            ChainTracker {
+                select_chain,
+                client,
+                data_to_propose: data_to_propose.clone(),
+                min_block_num,
+                max_block_num,
+                best_block_in_session: None,
+                prev_chain_info: None,
+                _phantom: PhantomData,
+            },
+            DataProvider {
+                data_to_propose,
+                metrics,
+            },
+        )
+    }
+
+    fn update_data(&mut self) {
+        // We read the best_block_in_session field and the highest_finalized block from the client and compute
+        // the corresponding `AlephData<B>` in `data_to_propose` fort AlephBFT. To not recompute this many
+        // times we remember these "inputs" in `prev_chain_info` and upon match we leave the old value
+        // of `data_to_propose` unaffected.
+
+        if let Some(best_block) = &self.best_block_in_session {
+            let client_info = self.client.info();
+            let finalized_block =
+                BlockHashNum::<B>::new(client_info.finalized_hash, client_info.finalized_number);
+
+            if let Some(prev) = &self.prev_chain_info {
+                if prev.best_block_in_session == *best_block
+                    && prev.highest_finalized == finalized_block
+                {
+                    // This is exactly the same state that we processed last time in update_data.
+                    // No point in recomputing.
+                    return;
                 }
-                prev_best_hash = new_best_header.hash();
-                prev_best_number = *new_best_header.number();
             }
-            _ = &mut exit => {
-                debug!(target: "afa", "Task for refreshing best chain received exit signal. Terminating.");
+            // Update the info for the next call of update data.
+            self.prev_chain_info = Some(ChainInfoForData {
+                best_block_in_session: best_block.clone(),
+                highest_finalized: finalized_block.clone(),
+            });
+            let mut curr_block = best_block.clone();
+
+            if curr_block.num == finalized_block.num {
+                // We don't have anything to propose, we go ahead with an empty propasal.
+                *self.data_to_propose.lock() = AlephData::HeadProposal(AlephProposal::<B>::new(
```

Comment:
This should be `Empty` indeed, that's one of the reasons I actually introduced `Empty`, but forgot to update it.

The general problem with not having `Empty` is the following:
1. I want to keep the invariant that every proposal in session `k` is a block belonging to session `k`.
2. I want `get_data` to output data without blocking.

The invariant 1. is just something that makes sense and is reasonable. 
Otherwise, introducing `Empty` actually made the code much nicer in many places. If there is no new block to propose, then why would we propose an old block?

When it comes to saving on the size of `Data`, this was not at all my intention, and I don't think we save anything here.

Reply:
I would argue for keeping the invariant of always proposing _something_, since it would be easier to see what went wrong in some cases. This would require relaxing invariant 1. from your list to "...belonging to session `k` or the last block of session `k-1`", which I agree is somewhat ugly, but imo likely worth it.

On the other hand I am not particularly sure this is better, maybe I'm just attached to the assumption that every unit has the highest known block in it for no reason. Maybe we should talk about this more personally, I'll reach out.